### PR TITLE
Centralize mechanism data files

### PR DESCRIPTION
**Pull request by MTCam**
Uses a `MANIFEST.in` and `importlib-resources` to manage the CTI mechanism data files. Builds on #147.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ numpy
 pytest
 pyvisfile
 pymetis
+importlib-resources
```
Comment (inducer) [inline requirements.txt:7]:
```suggestion
importlib-resources>=1.1.0; python_version < '3.9'
```

File: setup.py [hunk 2]
```diff
@@ -58,7 +58,8 @@ def main():
               "dagrt>=2019.1",
               "grudge>=2015.1",
               "six>=1.8",
-              ])
+              ],
```
Comment (inducer) [inline setup.py:61]:
```suggestion
              "importlib-resources>=1.1.0; python_version < '3.9'",
              ],
```

File: test/test_eos.py [hunk 3]
```diff
@@ -89,8 +99,8 @@ def test_pyrometheus_mechanisms(ctx_factory, mechname, y0):
     nodes = thaw(actx, discr.nodes())
 
     # Hand-port chemistry works OK
-    # prometheus_mechanism = UIUCMechanism(actx.np)
-    sol = cantera.Solution(f"{mechname}.cti", "gas")
+    mech_path = str(mechdata / f"{mechname}.cti")
```
Comment (inducer) [inline test/test_eos.py:102]: This won't work if the import location is actually a zip file. Either copy to a temp file and give its name to cantera, or see if cantera will take a file-like. Similar below.

Comment (MTCam) [reply test/test_eos.py:102]: I do believe you, though would like to say I was following the [manual](https://importlib-resources.readthedocs.io/en/latest/using.html), down where it goes:
```
In general you never have to worry whether your package is on the file system or in a zip file, as the importlib_resources APIs hide those details from you. Sometimes though, you need a path to an actual file on the file system (...)
To support this, importlib_resources provides an API that will extract the resource from a zip file to a temporary file, and return the file system path to this temporary file as a pathlib.Path object.
```

Comment (inducer) [reply test/test_eos.py:102]: Oh TIL. Nvm in that case.

**Comment by inducer** [comment]
(My impression was that the package data would live under `mirgecom`. But I could be wrong on that.)

**Comment by inducer** [comment]
[This](https://github.com/inducer/pyopencl/blob/7dc035a8d857ef189ab321354002b4c7aa241071/setup.py#L257-L264) is the one bit of first-hand experience I have.

File: test/test_eos.py (hunk 3 above)
Comment (MTCam) [reply test/test_eos.py:102]: Not so fast, actually.

Comment (MTCam) [reply test/test_eos.py]: I think that there is something I am doing wrong with that path-to-file approach. This change (6fa2ca3, 37ee711) might actually be better in that it (I think) addresses your original concern, and it has the added value that it actually works.

**Comment by MTCam** [comment]
> [This](https://github.com/inducer/pyopencl/blob/7dc035a8d857ef189ab321354002b4c7aa241071/setup.py#L257-L264) is the one bit of first-hand experience I have.

I think I duplicated this. (9b3fb90)

**Comment by MTCam** [comment]
When we use this interface:
```python
    mech_file = mechdata / "uiuc.cti"
    with mech_file.open() as fp:
        mech_cti = fp.read()
```
... importing from a ZIP installation, `importlib_resources` creates a temporary file that is automatically cleaned up when closed.  Here we're reading all the text of the CTI file into a string which is later used to create the Cantera object with one of its available constructors.

File: MANIFEST.in [hunk 4]
```diff
@@ -0,0 +1 @@
+graft mechanisms
```
Comment (inducer) [inline MANIFEST.in:1]:
```suggestion
include mechanisms/*.cti
```
(`graft` has the potential to pick up unwanted files, such as backups)

File: test/test_eos.py [hunk 5]
```diff
@@ -54,8 +55,17 @@
...
+    # importlib.resources either doesn't exist or lacks the files()
+    # function, so use the PyPI version:
+    import importlib_resources
+else:
+    # importlib.resources has files(), so use that:
+    import importlib.resources as importlib_resources
+
 
 logger = logging.getLogger(__name__)
+mechdata = importlib_resources.files("mechanisms")
```
Comment (inducer) [inline test/test_eos.py:68]: We could package this up in a function, so we don't have to repeat it? `get_mechanism_path`?

File: examples/mixture-mpi.py [hunk 6]
```diff
@@ -55,8 +55,18 @@
...
+    # importlib.resources either doesn't exist or lacks the files()
+    # function, so use the PyPI version:
+    import importlib_resources
+else:
+    # importlib.resources has files(), so use that:
+    import importlib.resources as importlib_resources
+
 
 logger = logging.getLogger(__name__)
+mechdata = importlib_resources.files("mechanisms")
```
Comment (inducer) [inline examples/mixture-mpi.py:69]: I think this is ultimately mis-scoped: It effectively introduces a top-level package `mechanisms`, when IMO `mirgecom.mechanisms` would be more appropriate.

x-ref: https://github.com/illinois-ceesd/mirgecom/pull/207/files

File: setup.py [hunk 7]
```diff
@@ -58,7 +58,15 @@ def main():
               "dagrt>=2019.1",
               "grudge>=2015.1",
               "six>=1.8",
-              ])
+              "importlib-resources>=1.1.0; python_version < '3.9'",
+              ],
+
+          include_package_data=True,
+          package_data={
```
Comment (inducer) [inline setup.py:65]: IMO `mechanisms/__init__.py` shouldn't be necessary. There's not an `__init__.py` in pyopencl's package data either.

**Review by inducer — COMMENTED** [review]
Thanks!

File: MANIFEST.in (hunk 4 above)
Comment (MTCam) [reply MANIFEST.in]: bb15df6

File: setup.py (hunk 7 above)
Comment (MTCam) [reply setup.py:65]: 51f3378

File: MANIFEST.in (hunk 4 above)
Comment (MTCam) [reply MANIFEST.in]: This change seems to break the examples.

File: examples/mixture-mpi.py (hunk 6 above)
Comment (MTCam) [reply examples/mixture-mpi.py:69]: I'm trying to figure out how to make this happen, but not having much luck with it.

File: setup.py (hunk 7 above)
Comment (MTCam) [reply setup.py:65]: To be continued

**Comment by MTCam** [comment]
This is almost a whole different change set - sorry bout that.   Some things came back that we tried to remove (e.g. `graft` and `__init__`).  I think the (more important) structural/scoping issue is resolved in this latest set, however.

I'm not entirely happy with it.  I'd rather:
```
from mirgecom.mechanisms import get_mechnasim_cti 
```
And/or maybe the junk in `mechutil` could go in `mechanisms/__init__.py`?

File: test/test_eos.py (hunk 5 above)
Comment (MTCam) [reply test/test_eos.py]: 904b6bc  Yeah, I think this is much better.

File: examples/mixture-mpi.py (hunk 6 above)
Comment (MTCam) [reply examples/mixture-mpi.py]: 4742033  Ok, this is much closer to what we want here, but I think there's still some un-needed infrastructure stuff.

File: MANIFEST.in (hunk 4 above)
Comment (MTCam) [reply MANIFEST.in]: No it doesn't.  09f5666

File: setup.py (hunk 7 above)
Comment (MTCam) [reply setup.py]: I'm afraid I've gotten even deeper into `__init__.py`.  Is my use of it still in error? (0b6cb49)

**Comment by MTCam** [comment]
Note: We need to be careful with the return type documentation for the `importlib` function.   Apparently the return type depends on whether you end up with the built-in `importlib` vs. the back-port to python < 3.9.  In both packages the returned object is _used_ exactly the same. The both _look like_ and duplicate features of an `os.path` object, but they are something more.

**Comment by MTCam** [comment]
This work is superseded by #232.

**Pull request closed without merging**